### ISSUE TITLE
WebSocket instability

### ISSUE THREAD

**Issue by jflatow** [denoland/deno#22333]
Version: at least Deno 1.39.1 through 1.40.3

This is related to #17283 and probably other issues, but recently WebSockets seem to have gotten more unstable. Simply refreshing pages or navigating in a browser which is connected via a WebSocket to a Deno server can lead to various crashes of the form: `Upgrade response was not returned from callback`. This seems like a regression as it was previously relatively stable (except when using TLS).

**Comment by e3dio** [comment]
This error `Upgrade response was not returned from callback` https://github.com/denoland/deno/blob/main/ext/http/00_serve.js#L483 happens for a specific case: using `Deno.serve` and `Deno.upgradeWebSocket` and then not returning the same response that was returned by `Deno.upgradeWebSocket`. Does your upgrade route return the correct response object in all cases ?

```js
const handler = req => {
   const { socket, response } = Deno.upgradeWebSocket(req);
   return response;
}
```

I don't know about it crashing the server though you would need to provide more info, it would just close the client connection

**Comment by bartlomieju** [comment]
CC @mmastrac @littledivy

**Comment by littledivy** [comment]
We would need some code (atleast the `serve` handler) to investigate this. 

The error occurs when the response returned from the `serve` handler is not the same one returned by `Deno.upgradeWebSocket`. This is intended behaviour.

**Comment by jflatow** [comment]
> This error `Upgrade response was not returned from callback` https://github.com/denoland/deno/blob/main/ext/http/00_serve.js#L483 happens for a specific case: using `Deno.serve` and `Deno.upgradeWebSocket` and then not returning the same response that was returned by `Deno.upgradeWebSocket`. Does your upgrade route return the correct response object in all cases ?
> 
> ```js
> const handler = req => {
>    const { socket, response } = Deno.upgradeWebSocket(req);
>    return response;
> }
> ```
> 
> I don't know about it crashing the server though you would need to provide more info, it should just close the client connection if correct response object is not returned

I have no branches and always return this response, and the code I have hasn't changed and was previously stable, so I'm not sure (and it does fully crash now). I will work on producing a minimal example, but I don't think you should just dismiss this.

**Comment by e3dio** [comment]
> I have no branches and always return this response

I do see an issue/bug with the new Deno.serve() and websocket route: if the route throws an error before response gets returned, instead of the server returning the generated 500 error response, the server stops accepting requests and process exits if nothing else is running, this needs to be fixed. I'm guessing your route threw an error at some point before returning response

```js
const port = 3003;

const server = Deno.serve({ port }, req => {
	const { socket, response } = Deno.upgradeWebSocket(req);
	throw 'throw error test';
	return response;
});

const ws = new WebSocket(`ws://127.0.0.1:${port}`);

server.finished.then(() => console.log('server exited'));
```
> throw error test
Upgrade response was not returned from callback
server exited

https://github.com/denoland/deno/blob/main/ext/http/00_serve.js#L468-L487

**Comment by mmastrac** [comment]
@e3dio Great catch. I think if a connection has been upgraded and the handler throws we should avoid calling the onError handler and treat it as success.

**Comment by e3dio** [comment]
It's tricky because the connection has been 'upgraded' but the websocket doesn't finishing opening server-side until the response is successfully returned
https://github.com/denoland/deno/blob/main/ext/http/00_serve.js#L486 
https://github.com/denoland/deno/blob/main/ext/http/00_serve.js#L207-L222
If an error is thrown before response is returned, it's likely the user would prefer the connection to fail and close rather than continue opening websocket

**Comment by e3dio** [comment]
I'm wondering if it would be simpler to wait until response is returned from callback before calling [op_http_upgrade_websocket_next](https://github.com/denoland/deno/blob/main/ext/http/00_serve.js#L195C25-L195C55)
That way the upgrade happens in 1 step at same time, instead of weboscket existing in half-open state. Currently websocket is open according to client, but not officially open on server until response is returned. There could be a delay that causes issues

**Comment by e3dio** [comment]
I see this [goAhead](https://github.com/denoland/deno/blob/main/ext/http/00_serve.js#L207) gets resolved before the response is returned so it does fully open the websocket before response returned. I don't see where this is getting resolved though, it's not getting called from the only place I see it can get called which is after the returned response 
https://github.com/denoland/deno/blob/main/ext/http/00_serve.js#L207

**Comment by mmastrac** [comment]
Good point. The `goAhead` signal is triggered when the response is returned. This ensures the WebSocket cannot be used until the response transaction has completed.

This is mostly for compatibility with the existing Deno.serveHttp implementation. But since the WebSocket is open-but-technically-inert until the response is returned, we can probably just abort the WebSocket if the response handler throws.

**Comment by e3dio** [comment]
> But since the WebSocket is open-but-technically-inert until the response is returned

`goAhead` is getting called before the response is returned, it must be called from some rust code that sends the upgrade response, I don't see where yet. So technically the websocket can be fully open and ready to use before the response object is returned in callback, I tested by waiting before returning response and it works.

A quick fix for current issue is to close websocket connection if error in request callback, and not stop server from accepting new requests. Whether the websocket should officially be upgraded during `Deno.upgradeWebSocket(req)` call or after response is returned from callback is a different question. Probably makes more sense to wait until response is returned, then it can follow the same request/response/error pipeline as a standard request

**Comment by e3dio** [comment]
Ok I think I see mistake in code, this should be `await goAhead.promise` not `await goAhead` https://github.com/denoland/deno/blob/main/ext/http/00_serve.js#L207
https://github.com/denoland/deno/blob/main/ext/web/06_streams.js#L125

So it doesn't actually wait for the promise

**Comment by e3dio** [comment]
So you have mistake in code there that triggered WS open early. But even if you fix that and `await goAhead.promise` now you have half-open websocket, open on client side but not officially on server. Ideally connection upgrade should happen after response is returned. You able to fix all these things or you need some help with that

**Comment by mmastrac** [comment]
I'll PR a fix for the goAhead bug to start -- that's definitely not behaving as we expect.

**Comment by pagoru** [comment]
This is not a bug, this error happens when you `upgradeWebSocket` the request and then, return a `new response`. The response needs to be from the socket, because you already upgraded the request to a socket, so it makes sense to throw that error.

**Comment by jflatow** [comment]
> This is not a bug, this error happens when you `upgradeWebSocket` the request and then, return a `new response`. The response needs to be from the socket, because you already upgraded the request to a socket, so it makes sense to throw that error.

No, see above, this is/was definitely a bug since it was happening when the upgrade response was being returned correctly.

**Comment by pagoru** [comment]
> > This is not a bug, this error happens when you `upgradeWebSocket` the request and then, return a `new response`. The response needs to be from the socket, because you already upgraded the request to a socket, so it makes sense to throw that error.
> 
> No, see above, this is/was definitely a bug since it was happening when the upgrade response was being returned correctly.

Ups, sorry! 
I had a similar problem, but was my fault because I was modifying the response.

If anyone happens to reach this and are modifying the response or sending a new response, Don't do it!

**Comment by MiguelRipoll23** [comment]
Using hono:

```
this.app.get("/websocket", (c) => {
  const { response } = Deno.upgradeWebSocket(c.req.raw);

  return response;
});
```

I see on my console:

> Upgrade response was not returned from callback

I tried to use WebSocket Helper with no luck.

**Comment by manu-xo5** [comment]
thanks @MiguelRipoll23

**Comment by bartlomieju** [comment]
Closing as not reproducible on current versions. The HTTP/WebSocket server stack was substantially rewritten (hyper 1.x migration) since the affected 1.39–1.40 releases.

Since this was filed without a repro (`needs info`), I built stress tests mimicking the reported trigger (browsers refreshing/navigating, i.e. abrupt disconnects) on current `main`:

**Plaintext WS** — 2000 connections dropped abruptly at randomized points (partial handshake, full handshake then drop, drop right after the 101, bare connect-and-drop), in waves of 100:
- 0 `Upgrade response was not returned from callback` logs
- No panics; server stayed alive
- Clean teardown: `upgrades == opens == closes == errors` (every established-then-dropped socket fired `onclose`, no leaks)
- A normal WS echo still worked afterward

**WSS (TLS)** — 800 connect-then-immediately-close churn:
- 0 error logs, no panics, server alive
- `upgrades=800, opens=800, errors=0`, clean closes

No crashes or instability under either transport. If this still reproduces on a current Deno release, please reopen with a minimal server + client repro and the exact version, and we'll take another look.
